Fast Confirmation Rule (#4747)
ethereum/consensus-specs: specs/phase0/fast-confirmation.md

```python
def get_ancestor_roots(store: Store, block_root: Root, terminal_root: Root) -> Sequence[Root]:
    """
    Return a list of ancestors of ``block_root`` inclusive until ``terminal_root`` exclusive.
    """
    root = block_root
    ancestor_roots: list[Root] = []
    while store.blocks[root].slot > store.blocks[terminal_root].slot:
        ancestor_roots.insert(0, root)
        root = store.blocks[root].parent_root

        # Return when terminal_root is reached
        if root == terminal_root:
            return ancestor_roots

    # Return empty list if terminal_root is not in the chain of block_root
    return []
```

#### State helpers

This section encapsulates reads from different beacon states used by the Fast
Confirmation Rule.

Implementations MAY override the logic of each of these functions but the
semantics MUST be preserved.

##### `get_slot_committee`

*Note:* This function returns the committee for a specific slot. It MUST support
committees of epochs starting from `current_epoch - 2`.

```python
def get_slot_committee(store: Store, slot: Slot) -> Set[ValidatorIndex]:
    """
    Return participants of all committees in ``slot``.
    """

>     head = get_head(store)

Reviewer: Should this be `get_pulled_up_head_state(store)` instead? I'm afraid the results of this could be incorrect around epoch boundaries.

Author: Epoch processing affects validator state but the committee shuffling computation uses validator set active at the epoch for which committees are computed, validator effective balance doesn’t affect shuffling too. So, the head and the pulled up head states should be equivalent in this computation.

            )
        )
    )
```

##### `is_full_validator_set_covered`

```python
def is_full_validator_set_covered(start_slot: Slot, end_slot: Slot) -> bool:
    """
    Return ``True`` if the range between ``start_slot`` and ``end_slot`` (inclusive of both) includes an entire epoch.
    """
    start_full_epoch = compute_epoch_at_slot(start_slot + (SLOTS_PER_EPOCH - 1))
    end_full_epoch = compute_epoch_at_slot(Slot(end_slot + 1))

    return start_full_epoch < end_full_epoch
```

##### `adjust_committee_weight_estimate_to_ensure_safety`

*Note*: This function adjusts the estimate of the weight of a committee for a
sequence of slots not covering a full epoch to ensure the safety of FCR with
high probability. See
https://gist.github.com/saltiniroberto/9ee53d29c33878d79417abb2b4468c20 for an
explanation of why this is required.

```python
def adjust_committee_weight_estimate_to_ensure_safety(estimate: Gwei) -> Gwei:
    """
    Return adjusted ``estimate`` of the weight of a committee for a sequence of slots not covering a full epoch.
    """
    ceil = (estimate + 999) // 1000
    return Gwei(ceil * (1000 + COMMITTEE_WEIGHT_ESTIMATION_ADJUSTMENT_FACTOR))

> ```

Reviewer: In this section, I believe we should replace "not covering a full epoch" with "for a sequence of slots spanning an epoch boundary that does not cover any full epoch. The sequence may be longer than SLOTS_PER_EPOCH." for more clarity. The gist tells me that is possible, but reading this note & docstring, I wouldn't have thought that.

> We want to estimate the committee weight in a range of slots starting in the previous epoch & ending in the current epoch. Say the range covers X slots in the current epoch & Y slots in the previous epoch.

Author: Done!

    """
    root = block_root
    ancestor_roots: list[Root] = []
    while store.blocks[root].slot > store.blocks[terminal_root].slot:
        ancestor_roots.insert(0, root)
        root = store.blocks[root].parent_root

        # Return when terminal_root is reached
        if root == terminal_root:
            return ancestor_roots

    # Return empty list if terminal_root is not in the chain of block_root
    return []
```

#### State helpers

This section encapsulates reads from different beacon states used by the Fast
Confirmation Rule.

Implementations MAY override the logic of each of these functions but the
semantics MUST be preserved.

##### `get_slot_committee`

*Note:* This function returns the committee for a specific slot. It MUST support
committees of epochs starting from `current_epoch - 2`.

```python
def get_slot_committee(store: Store, slot: Slot) -> Set[ValidatorIndex]:
    """
    Return participants of all committees in ``slot``.
    """
    head = get_head(store)
    shuffling_source = store.block_states[head]
    committees_count = get_committee_count_per_slot(shuffling_source, compute_epoch_at_slot(slot))

>     participants: set[ValidatorIndex] = set()

Reviewer 2: ```suggestion
    participants: Set[ValidatorIndex] = set()
```

Keep the same capital typing-style as in other areas of specs.

of slots are consistent with the `balance_source` validator set.

```python
def get_block_support_between_slots(
    store: Store,
    balance_source: BeaconState,
    block_root: Root,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return support of the block by validators assigned to slots
    between ``start_slot`` and ``end_slot`` (inclusive of both).
    """
    participants: Set[ValidatorIndex] = set()
    for slot in range(start_slot, end_slot + 1):
        participants.update(get_slot_committee(store, Slot(slot)))

    # Keep validators that were active at the balance_source epoch to be consistent
    # with get_total_active_balance() computation, also filter out slashed validators
    unslashed_and_active_indices = [
        i
        for i in participants
        if (
            not balance_source.validators[i].slashed
            and is_active_validator(balance_source.validators[i], get_current_epoch(balance_source))
        )
    ]

    return Gwei(
        sum(
            balance_source.validators[i].effective_balance
            for i in unslashed_and_active_indices
            # Check that validator has voted in the support of the block

>             # and has not been slashed

Reviewer 2: ```suggestion

```

Can be removed as we're already iterating over `unslashed` indices, that was already filtered out previously

Author: Not really, some validators might be unslashed on chain, but at the same time slashed off the chain. Unslashed in `unslashed_and_active_indices` filters out on chain slashings

        # Next, calculate the number of slots remaining in the end epoch
        remaining_slots_in_end_epoch = SLOTS_PER_EPOCH - num_slots_in_end_epoch
        # Then, calculate the number of slots in the start epoch
        num_slots_in_start_epoch = SLOTS_PER_EPOCH - compute_slots_since_epoch_start(start_slot)

        start_epoch_weight = committee_weight * num_slots_in_start_epoch
        end_epoch_weight = committee_weight * num_slots_in_end_epoch

        # A range that spans an epoch boundary, but does not span any full epoch
        # needs pro-rata calculation, see https://gist.github.com/saltiniroberto/9ee53d29c33878d79417abb2b4468c20
        # start_epoch_weight_pro_rated = start_epoch_weight * (1 - num_slots_in_end_epoch / SLOTS_PER_EPOCH)
        start_epoch_weight_pro_rated = (
            start_epoch_weight // SLOTS_PER_EPOCH * remaining_slots_in_end_epoch
        )

        return adjust_committee_weight_estimate_to_ensure_safety(
            Gwei(start_epoch_weight_pro_rated + end_epoch_weight)
        )
```

##### `get_equivocation_score`

*Notes:*

For simplicity, this function does not seek `balance_source` for slashed

> validators as it is very unlikely that those validators are not already in

Reviewer 2: ```suggestion
validators as those validators are very likely already in
```

simplify the double negation

`store.equivocating_indices`.

Due to the algorithm logic, maximum distance between `balance_source` and
`start_slot` or `end_slot` is two epochs, which is less than
`MAX_SEED_LOOKAHEAD`. Therefore, participants of the committees from that span
of slots are consistent with the `balance_source` validator set.

```python
def get_equivocation_score(
    store: Store,
    balance_source: BeaconState,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return total weight of equivocating participants of all committees
    in the slots between ``start_slot`` and ``end_slot`` (inclusive of both).
    """

>     committee_indices: set[ValidatorIndex] = set()

Reviewer 2: ```suggestion
    committee_indices: Set[ValidatorIndex] = set()
```

Keep the same capital typing-style as in other areas of specs.

        return Gwei(0)

    # Discount votes supporting the parent block if they are from the committees of empty slots
    parent_support_in_empty_slots = get_block_support_between_slots(
        store,
        balance_source,
        block.parent_root,
        Slot(parent_block.slot + 1),
        Slot(block.slot - 1),
    )
    # Adversarial weight is not discounted
    adversarial_weight = compute_adversarial_weight(
        store, balance_source, Slot(parent_block.slot + 1), Slot(block.slot - 1)
    )
    if parent_support_in_empty_slots > adversarial_weight:
        return parent_support_in_empty_slots - adversarial_weight
    else:
        return Gwei(0)
```

##### `get_support_discount`

```python
def get_support_discount(store: Store, balance_source: BeaconState, block_root: Root) -> Gwei:
    """
    Return weight that can be discounted during the safety threshold computation for the block.
    """
    return compute_empty_slot_support_discount(store, balance_source, block_root)
```

##### `compute_safety_threshold`

```python
def compute_safety_threshold(store: Store, block_root: Root, balance_source: BeaconState) -> Gwei:
    """

>     Compute the LMD_GHOST safety threshold for ``block_root``.

Reviewer 2: ```suggestion
    Compute the LMD-GHOST safety threshold for ``block_root``.
```

Other areas says `LMD-GHOST`, instead of `LMD_GHOST`
Use hyphen in place of underscore

At a high level the computation checks whether the actual score of the block
outweighs potential score of any block conflicting with it considering total
weight of the committees and maximal adversarial weight.

If this check passes the block is deemed LMD-GHOST safe, but it is not enough to
say that the block will remain canonical. To ensure the latter, each ancestor of
the block would also have to pass this check.

More details on this check can be found in the
[paper](https://arxiv.org/abs/2405.00549).

This function **MUST** return `False` if `block_root` status is **not** `VALID`
according to the [Optimistic sync](../../sync/optimistic.md) specification.

```python
def is_one_confirmed(store: Store, balance_source: BeaconState, block_root: Root) -> bool:
    """
    Return ``True`` if and only if the block is LMD-GHOST safe.
    """
    support = get_attestation_score(store, block_root, balance_source)
    safety_threshold = compute_safety_threshold(store, block_root, balance_source)
    return support > safety_threshold
```

##### `is_confirmed_chain_safe`

*Notes*:

This function should be called at the start of each epoch to ensure that the
confirmed chain starting from

> `store.current_epoch_observed_justified_checkpoint.root` remains LMD-GHOST safe.

Reviewer 2: Should this one also be `fcr_store` ? If it was missed by the previous commit to change it

Author: Well spotted, fixed 👍

    adversarial_weight = compute_adversarial_weight(
        store, balance_source, compute_start_slot_at_epoch(current_epoch), Slot(current_slot - 1)
    )

    # Compute min honest FFG support
    min_honest_ffg_support = ffg_support_for_checkpoint - min(
        adversarial_weight, ffg_support_for_checkpoint
    )

    return Gwei(min_honest_ffg_support + remaining_honest_ffg_weight)
```

##### `will_no_conflicting_checkpoint_be_justified`

*Note:* This function assumes that all honest validators will be voting in
support of the current epoch target starting from the current moment in time.

```python
def will_no_conflicting_checkpoint_be_justified(store: Store) -> bool:
    """
    Return ``True`` if and only if no checkpoint conflicting with the current target can ever be justified.
    """

    # If the target is unrealized justified then no conflicting checkpoint can be justified
    if get_current_target(store) == store.unrealized_justified_checkpoint:
        return True

    state = get_pulled_up_head_state(store)
    total_active_balance = get_total_active_balance(state)
    honest_ffg_support = compute_honest_ffg_support_for_current_target(store)

>     return 3 * honest_ffg_support > 1 * total_active_balance

Reviewer 2: ```suggestion
    return 3 * honest_ffg_support > total_active_balance
```

Redundant multiplier, unless it was explicitly for clarity

Author: Yeah, the main goal was highlighting difference with `will_current_target_be_justified` which operates with a threshold at 2/3

    fcr_store: FastConfirmationStore, latest_confirmed_root: Root
) -> Root:
    """
    Return the most recent confirmed block in the suffix of the canonical chain
    starting from ``latest_confirmed_root``.
    """
    store = fcr_store.store
    head = get_head(store)
    current_epoch = get_current_store_epoch(store)
    confirmed_root = latest_confirmed_root

    if (
        get_block_epoch(store, confirmed_root) + 1 == current_epoch
        and get_voting_source(store, fcr_store.previous_slot_head).epoch + 2 >= current_epoch
        and (
            is_start_slot_at_epoch(get_current_slot(store))
            or (
                will_no_conflicting_checkpoint_be_justified(store)
                and (
                    store.unrealized_justifications[fcr_store.previous_slot_head].epoch + 1
                    >= current_epoch
                    or store.unrealized_justifications[head].epoch + 1 >= current_epoch
                )
            )
        )
    ):
        # Get suffix of the canonical chain
        canonical_roots = get_ancestor_roots(store, head, confirmed_root)

        # Starting with the child of the latest_confirmed_root

>         # move towards the head in attempt to advance confirmed block

Reviewer 2: ```suggestion
        # move towards the head in attempt to advance the confirmed block
```

add the missing article

                # To confirm blocks from the current epoch ensure that
                # current epoch target will be justified
                if not will_current_target_be_justified(store):
                    break

            if not is_one_confirmed(store, get_current_balance_source(fcr_store), block_root):
                break

            tentative_confirmed_root = block_root

        # The tentative_confirmed_root can only be confirmed
        # if it is for sure not going to be reorged out in either the current or next epoch.
        if get_block_epoch(store, tentative_confirmed_root) == current_epoch or (
            get_voting_source(store, tentative_confirmed_root).epoch + 2 >= current_epoch
            and (
                is_start_slot_at_epoch(get_current_slot(store))
                or will_no_conflicting_checkpoint_be_justified(store)
            )
        ):
            confirmed_root = tentative_confirmed_root

    return confirmed_root
```

#### `get_latest_confirmed`

*Notes:*

This function executes the FCR algorithm which takes the following sequence of
actions:

1. Check if the `store.confirmed_root` belongs to the canonical chain and is not

>    older than the previous epoch.

Reviewer 3: Shouldn't these all be `fcr_store` instead of `store`?

Author: Good catch! updated in this and all other places

   `start_slot`, but votes for `block_root` in `end_slot + 1`.

In both cases the support would count a vote outside of the
`[start_slot; end_slot]` range. This inaccuracy is acceptable as it doesn't
affect safety.

Due to the algorithm logic, maximum distance between `balance_source` and
shuffling is two epochs which is less than `MAX_SEED_LOOKAHEAD` and thus the
balances are always consistent with the shuffling.

```python
def get_block_support_between_slots(
    store: Store,
    balance_source: BeaconState,
    block_root: Root,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return support of the block by validators assigned to slots
    between ``start_slot`` and ``end_slot`` (inclusive of both).
    """
    participants: set[ValidatorIndex] = set()
    for slot in range(start_slot, end_slot + 1):
        participants.update(get_slot_committee(store, Slot(slot)))

    # Keep validators that were active at the balance_source epoch to be consistent
    # with get_total_active_balance() computation, also filter out slashed validators
    unslashed_and_active_indices = [
        i
        for i in participants
        if (

>             i < len(balance_source.validators)

Author: It depends on the distance between `balance_source` and `shuffling_source`. As per the algorithm it should not be greater than 2 epochs which guarantees that no new validator emerges in the committee between `balance_source` and `shuffling_source`, at least because of `MAX_SEED_LOOKAHEAD`.

If that distance would be greater because for instance algorithm could use a more distant balance source (which is theoretically possible and shouldn’t break safety cc @saltiniroberto) then that could happen.

Do you have a strong preference to remove this safe guard considering the above?

>             i < len(balance_source.validators)

Reviewer: Thanks for the extra information. I don't have a strong preference to remove it, but generally I'm opposed to checks for "impossible" situations. Alternatively, you could define a constraint that the distance is less than or equal to `MAX_SEED_LOOKAHEAD+1`, then remove the safe guards. Sort of like:

https://github.com/ethereum/consensus-specs/blob/bfbef99d7dce0c56a2032cc737fb63163425a3d9/tests/core/pyspec/eth_consensus_specs/test/phase0/unittests/test_config_invariants.py#L71-L80

>             i < len(balance_source.validators)

Author: I have remove this check and added a comment on that this is safe and the requested committee participants are consistent with `balance_source` validator set. Same is applied to `get_equivocation_score` function

    # with an underflow guard
    if support_discount < maximum_support + proposer_score + 2 * adversarial_weight:
        return (maximum_support + proposer_score + 2 * adversarial_weight - support_discount) // 2
    else:
        return Gwei(0)
```

##### `is_one_confirmed`

*Notes:*

This function checks if a single block is LMD-GHOST safe by computing LMD-GHOST
safety indicator and comparing its value to the safety threshold.

At a high level the computation checks whether the actual score of the block
outweighs potential score of any block conflicting with it considering total
weight of the committees and maximal adversarial weight.

If this check passes the block is deemed LMD-GHOST safe, but it's not enough to
say that the block will remain canonical. To ensure the latter, each ancestor of
the block would also have to pass this check.

More details on this check can be found in the
[paper](https://arxiv.org/abs/2405.00549).

This function **MUST** return `False` if `block_root` status is **not** `VALID`
according to the [Optimistic sync](../../sync/optimistic.md) specification.

```python

> def is_one_confirmed(store: Store, balance_source: BeaconState, block_root: Root) -> bool:

Author: This name is in the paper as well but I am open to change it to something like `is_block_lmd_confirmed`, cc @saltiniroberto. Although, I believe implementations already have `is_one_confirmed` function name I am not sure if it’s a problem to adjust the name in implementations.

> def is_one_confirmed(store: Store, balance_source: BeaconState, block_root: Root) -> bool:

Reviewer: Hmm I'm not going to push to get this changed. If it's too much of a headache, don't worry about it. I just found the name a little confusing.

      - [`get_current_target_score`](#get_current_target_score)
      - [`compute_honest_ffg_support_for_current_target`](#compute_honest_ffg_support_for_current_target)
      - [`will_no_conflicting_checkpoint_be_justified`](#will_no_conflicting_checkpoint_be_justified)
      - [`will_current_target_be_justified`](#will_current_target_be_justified)
    - [`update_fast_confirmation_variables`](#update_fast_confirmation_variables)
    - [`find_latest_confirmed_descendant`](#find_latest_confirmed_descendant)
    - [`get_latest_confirmed`](#get_latest_confirmed)
  - [Handlers](#handlers)
    - [`on_slot_start_after_past_attestations_applied`](#on_slot_start_after_past_attestations_applied)

<!-- mdformat-toc end -->

## Introduction

This document specifies a fast block confirmation rule (a.k.a. FCR) for the
Ethereum protocol.

The research paper for this rule can be found
[here](https://arxiv.org/abs/2405.00549).

A shorter explainer is available
[here](https://www.overleaf.com/project/691b4629fb781aeb8efdb20f).

This rule makes the following network synchrony assumption: starting from the
current slot, attestations created by honest validators in any slot are received

> by the end of that slot. Consequently, this rule provides confirmations to users

Reviewer 3: That end goal also comes with 3SF, I guess 😅 There is still a period where attestations may be late. It doesn't mean that it is a problem for FCR, just need to be aware.

```

#### LMD-GHOST helpers

##### `get_block_support_between_slots`

*Notes:*

Due to the algorithm logic, maximum distance between `balance_source` and
shuffling is two epochs which is less than `MAX_SEED_LOOKAHEAD` and thus the
balances are always consistent with the shuffling.

```python
def get_block_support_between_slots(
    store: Store,
    balance_source: BeaconState,
    block_root: Root,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return support of the block between ``start_slot`` and ``end_slot`` (inclusive of both).
    """
    participants: set[ValidatorIndex] = set()
    for slot in range(start_slot, end_slot + 1):
        participants.update(get_slot_committee(store, Slot(slot)))

    return Gwei(
        sum(
            balance_source.validators[i].effective_balance
            for i in participants
            # Check that validator is in the balance source set,
            # has voted in the support of the block
            # and hasn't been slashed.
            if (
                i < len(balance_source.validators)
                and i in store.latest_messages

>                 and store.latest_messages[i].root == block_root

Reviewer 3: OK, and just to double-check: That this function gives inaccurate results is okay?

                and i not in store.equivocating_indices

>                 and not balance_source.validators[i].slashed

Reviewer 3: Hmm, getting access to historical slashings is tricky. Would it be good enough to take the current head state for the slashed lookup? Or, alternatively, also feed proposer slashings and all slashed validators from the startup state into store.equivocating_indices. Or is there any real reason why equivocating_indices on one hand tracks across all branches (including non-canonical), but specifically limits itself to attesterslashings that happened since beacon node startup?

            )
        )
    )
```

##### `is_full_validator_set_covered`

```python
def is_full_validator_set_covered(start_slot: Slot, end_slot: Slot) -> bool:
    """
    Return ``True`` if the range between ``start_slot`` and ``end_slot`` (inclusive of both) includes an entire epoch.
    """
    start_full_epoch = compute_epoch_at_slot(start_slot + (SLOTS_PER_EPOCH - 1))
    end_full_epoch = compute_epoch_at_slot(Slot(end_slot + 1))

    return start_full_epoch < end_full_epoch
```

##### `adjust_committee_weight_estimate_to_ensure_safety`

*Notes*:

This function adjust the estimate of the weight of a committee for a sequence of
slots not covering a full epoch to ensure the safety of FCR with high
probability.

See https://gist.github.com/saltiniroberto/9ee53d29c33878d79417abb2b4468c20 for
an explanation of why this is required.

```python
def adjust_committee_weight_estimate_to_ensure_safety(estimate: Gwei) -> Gwei:
    """
    Return adjusted ``estimate`` of the weight of a committee for a sequence of slots not covering a full epoch.
    """

>     return Gwei(estimate // 1000 * (1000 + COMMITTEE_WEIGHT_ESTIMATION_ADJUSTMENT_FACTOR))

Reviewer 4: do you want to take the ceil instead of the floor here? I think the latex explainer rounds up

Author: Resolved by https://github.com/mkalinin/eth2.0-specs/pull/35

```

##### `estimate_committee_weight_between_slots`

```python
def estimate_committee_weight_between_slots(

>     state: BeaconState, start_slot: Slot, end_slot: Slot

Author: Addressed in https://github.com/mkalinin/eth2.0-specs/pull/23

    start_epoch = compute_epoch_at_slot(start_slot)
    end_epoch = compute_epoch_at_slot(end_slot)

    # Sanity check
    if start_slot > end_slot:
        return Gwei(0)

    # If an entire epoch is covered by the range, return the total active balance
    if is_full_validator_set_covered(start_slot, end_slot):
        return total_active_balance

    if start_epoch == end_epoch:
        return total_active_balance // SLOTS_PER_EPOCH * (end_slot - start_slot + 1)
    else:
        # First, calculate the number of committees in the end epoch
        num_slots_in_end_epoch = compute_slots_since_epoch_start(end_slot) + 1
        # Next, calculate the number of slots remaining in the end epoch
        remaining_slots_in_end_epoch = SLOTS_PER_EPOCH - num_slots_in_end_epoch
        # Then, calculate the number of slots in the start epoch
        num_slots_in_start_epoch = SLOTS_PER_EPOCH - compute_slots_since_epoch_start(start_slot)

        end_epoch_weight_estimate = total_active_balance // SLOTS_PER_EPOCH * num_slots_in_end_epoch
        start_epoch_weight_estimate = (
            total_active_balance
            // SLOTS_PER_EPOCH

>             // SLOTS_PER_EPOCH

Author: Addressed in https://github.com/mkalinin/eth2.0-specs/pull/23

            * num_slots_in_start_epoch
            * remaining_slots_in_end_epoch
        )

        # A range that spans an epoch boundary, but does not span any full epoch
        # needs pro-rata calculation
        return adjust_committee_weight_estimate_to_ensure_safety(
            Gwei(start_epoch_weight_estimate + end_epoch_weight_estimate)
        )
```

##### `get_equivocation_score`

*Note:* For simplicity, this function doesn't seek `balance_source` for slashed
validators as it is very unlikely that those validators aren't already in

> `store.equivocating_indices`.

Reviewer 3: Only attesterslashings that happened since the last node startup end up in equivocating_indices. proposerslashings and pre-existing slashings don't.

```python
def get_equivocation_score(
    store: Store,
    balance_source: BeaconState,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return total weight of equivocating participants of all committees
    in the slots between ``start_slot`` and ``end_slot`` (inclusive of both).
    """
    committee_indices: set[ValidatorIndex] = set()
    for slot in range(start_slot, end_slot + 1):

>         committee_indices.update(get_slot_committee(store, Slot(slot)))

Reviewer 3: I don't think it's necessary. Impls don't match this closely anyway.

In Nimbus, we don't even have `equivocating_indices`. We just pretend that the validator voted at FAR_FUTURE_SLOT. That implicitly disallows it to vote ever again.

The expensive part here is `get_slot_committee`. If one caches that value, it has to be cached before knowing whether at time of use, equivocating indices will be present.

Also, the FCR should not meaningfully become slower as soon as there is any attester slashing.

        # as then this block will always be canonical in this case.
        start_root = store.current_epoch_observed_justified_checkpoint.root
    else:
        # Limit reconfirmation to the checkpoint block
        # as if it's successful, reconfirmation of the ancestors is implied.
        checkpoint = get_checkpoint_for_block(store, confirmed_root, Epoch(current_epoch - 1))
        start_root = store.blocks[checkpoint.root].parent_root

    # Run is_one_confirmed for each block in the confirmed chain with the previous epoch balance source.
    chain_roots = get_ancestor_roots(store, confirmed_root, start_root)
    return all(
        is_one_confirmed(store, get_previous_balance_source(store), root) for root in chain_roots
    )
```

#### FFG helpers

##### `get_current_target_score`

*Notes:*

This function uses LMD-GHOST votes to estimate the FFG support of the current
epoch target. Due to the way the computation happens, it MUST be used no later
than the end of the current epoch.

```python
def get_current_target_score(store: Store) -> Gwei:
    """
    Return the estimate of FFG support of the current epoch target by using LMD-GHOST votes.
    """

>     target = get_current_target(store)

Reviewer 3: Implementers note: 
This function returns the regular fork choice weight of `target`, minus the proposer boost.
If the proposer boosted block is a descendant of `target`, have to discount it.
Likely cheaper than passing all the validators around.

Reviewer 3: Hmm, tiny difference would be that _proposer_ slashings not yet present in `target_state` that happened later would also be ignored. @mkalinin is that a problem if these proposer slashings are taken from the current head state instead of the target state?

Given it says "estimate" in the function description, maybe okay?

Reviewer 3: With, "regular fork choice weight", it still needs to be recomputed with the target's effective_balance of course, but one doesn't need to cache the slashings at the epoch start. The effective_balance don't change during the epoch, so can just look into the head state for those.

    target_state = get_current_target_state(store)
    unslashed_and_active_indices = [
        i
        for i in get_active_validator_indices(target_state, get_current_epoch(target_state))
        if not target_state.validators[i].slashed
    ]
    return Gwei(
        sum(
            target_state.validators[i].effective_balance
            for i in unslashed_and_active_indices
            if (
                i in store.latest_messages
                and i not in store.equivocating_indices
                and target
                == get_checkpoint_for_block(
                    store,
                    store.latest_messages[i].root,

>                     get_latest_message_epoch(store.latest_messages[i]),

Reviewer 3: Although, `get_current_epoch(target_state)` is the same as `get_current_epoch(store)`. So, it wouldn't change results.

    honest_ffg_support = compute_honest_ffg_support_for_current_target(store)
    return 3 * honest_ffg_support >= 1 * total_active_balance
```

##### `will_current_target_be_justified`

*Note:* This function assumes that all honest validators will be voting in
support of the current epoch target starting from the current moment in time.

```python
def will_current_target_be_justified(store: Store) -> bool:
    """
    Return ``True`` if and only if the current target will eventually be justified.
    """
    target_state = get_current_target_state(store)
    total_active_balance = get_total_active_balance(target_state)
    honest_ffg_support = compute_honest_ffg_support_for_current_target(store)
    return 3 * honest_ffg_support >= 2 * total_active_balance
```

#### `update_fast_confirmation_variables`

*Note:* This function updates variables used by the fast confirmation rule.

```python
def update_fast_confirmation_variables(store: Store) -> None:
    # Update prev and curr slot head.
    store.previous_slot_head = store.current_slot_head
    store.current_slot_head = get_head(store)

    # Update observed justified checkpoint at the beginning of the last slot of an epoch.

>     if is_start_slot_at_epoch(Slot(get_current_slot(store) + 1)):

Reviewer 3: Could the same be achieved by:

1. Add `update_fast_confirmation_variables` to the end of [`on_tick_per_slot`](https://github.com/ethereum/consensus-specs/blob/v1.7.0-alpha.2/specs/phase0/fork-choice.md#on_tick_per_slot)
2. In `update_fast_confirmation_variables` on the line of this comment, check for `get_current_slot(store)` instead of `get_current_slot(store) + 1` (as [`on_tick_per_slot`](https://github.com/ethereum/consensus-specs/blob/v1.7.0-alpha.2/specs/phase0/fork-choice.md#on_tick_per_slot) already advanced to the new epoch)
3. In `update_fast_confirmation_variables`, use `store.justified_checkpoint` instead of `store.unrealized_justified_checkpoint`, as unrealized chain tips have just been pulled by [`on_tick_per_slot`](https://github.com/ethereum/consensus-specs/blob/v1.7.0-alpha.2/specs/phase0/fork-choice.md#on_tick_per_slot).

That way:
- Attestations with slot-into-epoch 31 were not processed, because attestations can only affect the fork choice of subsequent slots. So, this is the same as today, where `update_fast_confirmation_variables` is called at start of slot 31, and only atts up through ≤ 30 are processed.
- Any impact from a block during slot 31's [`compute_pulled_up_tip`](https://github.com/ethereum/consensus-specs/blob/v1.7.0-alpha.2/specs/phase0/fork-choice.md#on_block) is taken into account. Meaning, that for FCR the exact same justified CP is used that the regular FC also uses.
- Fewer hooks! One can use the same `on_tick_per_slot` function that exists today, without worrying about "deferred attestations". And one can simply use the regular justified CP instead of dealing with unrealized data (which may or may not be easily available depending on the implementation).

Reviewer 3: Can't the same happen in the previous slot, where someone advances GU by sending block 30 solely to your node? so, you end up with a different GU at start of slot 31, than others?

Author: > Can't the same happen in the previous slot, where someone advances GU by sending block 30 solely to your node? so, you end up with a different GU at start of slot 31, than others?

In this case they will use an older GU for their FCR, this is fine. But what is important, it this case your node will use a CP that will be GJ or its ancestor for all honest nodes, so they will run their FC and attest to blocks based on that CP. Using different CPs in FCR if they are not conflicting is fine.

Author: > before the block for 31 is processed, because FCR only wants to consider information that was available up to 8s before epoch start, correct?

It can be after the block in slot 31 is processed but no later than 4s into a slot, so 8s are still left. In other words, calling FCR variables update after block is processed or at 4s mark should be fine

                    >= current_epoch
                    or store.unrealized_justifications[head].epoch + 1 >= current_epoch
                )
            )
        )
    ):
        # Get suffix of the canonical chain
        canonical_roots = get_ancestor_roots(store, head, confirmed_root)

        # Starting with the child of the latest_confirmed_root
        # move towards the head in attempt to advance confirmed block
        # and stop when the first unconfirmed descendant is encountered
        for block_root in canonical_roots:
            block_epoch = get_block_epoch(store, block_root)

            # If the current epoch is reached, exit the loop
            # as this code is meant to confirm blocks from the previous epoch
            if block_epoch == current_epoch:
                break

            # The algorithm can only rely on the previous head
            # if it is a descendant of the block that is attempted to be confirmed
            if not is_ancestor(store, store.previous_slot_head, block_root):
                break

            if not is_one_confirmed(store, get_current_balance_source(store), block_root):

>                 break

Reviewer 3: get_head is based on the greatest justified CP at time of call (including potential justifications during the epoch), while this one needs the historical greatest justified CP at the beginning of the epoch.

As `is_one_confirmed` calls `get_attestation_score`, that would mean that we have to re-run fork choice to get the historical weight based on the current balance source? Or store historical weights somewhere? Or is this another one of those things that doesn't have to be fully accurate, and just using the latest weight is fine in practice?

        and store.current_epoch_observed_justified_checkpoint.epoch + 1 == current_epoch
        and store.current_epoch_observed_justified_checkpoint
        == store.unrealized_justifications[head]
        and get_block_slot(store, confirmed_root)
        < get_block_slot(store, store.current_epoch_observed_justified_checkpoint.root)
    ):
        confirmed_root = store.current_epoch_observed_justified_checkpoint.root

    # Attempt to further advance the latest confirmed block.
    if get_block_epoch(store, confirmed_root) + 1 >= current_epoch:
        return find_latest_confirmed_descendant(store, confirmed_root)
    else:
        return confirmed_root
```

### Handlers

#### `on_slot_start_after_past_attestations_applied`

*Notes:*

This handler calls `update_fast_confirmation_variables` and then
`get_latest_confirmed` to update `store.confirmed_root` with the response of
that call.

Implementations MUST strictly follow the call sequence:

1. `update_fast_confirmation_variables`
2. `get_latest_confirmed`

Implementations MUST call `update_fast_confirmation_variables` at the start of a
slot after attestations from past slots have been applied. Otherwise, the

> synchrony assumption that the algorithm relies upon may not hold.

Reviewer 3: Can't we then.. bump it to the end of the slot? The GU can only change due to blocks, so the end result should be the same. Except, that now we also take in account any late attestations / attester slashings up to the end of the slot. Simplifies it as then the hook is simply just before the on_tick for the new slot.

    if get_block_epoch(store, block_root) > get_block_epoch(store, block.parent_root):
        # Use the first epoch slot as the start slot when crossing epoch boundary.
        start_slot = compute_start_slot_at_epoch(get_block_epoch(store, block_root))
        return compute_adversarial_weight(store, balance_source, start_slot, Slot(current_slot - 1))
    else:
        return compute_adversarial_weight(store, balance_source, block.slot, Slot(current_slot - 1))
```

##### `compute_empty_slot_support_discount`

```python
def compute_empty_slot_support_discount(
    store: Store, balance_source: BeaconState, block_root: Root
) -> Gwei:
    """
    Return weight that can be discounted during the safety threshold computation
    if there are empty slots preceding the block.
    """
    block = store.blocks[block_root]
    parent_block = store.blocks[block.parent_root]
    # No empty slot.
    if parent_block.slot + 1 == block.slot:
        return Gwei(0)

    # Discount votes supporting the parent block if they are from the committees of empty slots.
    parent_support_in_empty_slots = get_block_support_between_slots(
        store,
        balance_source,
        block.parent_root,

>         Slot(parent_block.slot + 1),

Reviewer 3: Impl note: This includes info from < previous_epoch (relative to current_slot)

    shuffling_source = store.block_states[head]
    committees_count = get_committee_count_per_slot(shuffling_source, compute_epoch_at_slot(slot))
    participants: set[ValidatorIndex] = set()
    for i in range(committees_count):
        participants.update(get_beacon_committee(shuffling_source, slot, CommitteeIndex(i)))
    return participants
```

##### `get_pulled_up_head_state`

```python
def get_pulled_up_head_state(store: Store) -> BeaconState:
    """
    Return the state of the head pulled up to the current epoch if needed.
    """
    head = get_head(store)
    head_state = store.block_states[head]
    if get_current_epoch(head_state) < get_current_store_epoch(store):
        pulled_up_state = copy(head_state)
        process_slots(pulled_up_state, compute_start_slot_at_epoch(get_current_store_epoch(store)))
        return pulled_up_state
    else:
        return head_state
```

##### `get_previous_balance_source`

*Notes:*

Reconfirmation is the only place where previous balance source is used.
Implementations MAY switch to the current epoch balance source after they run
reconfirmation and before the logic to confirm new blocks is called. In this

> case implementations will need to keep only a single balance source around.

Reviewer 3: Yep, previous_epoch_greatest_unrealized_checkpoint is now captured in slot 31, but the balance source is captured at slot 0. So, the comment now holds.

    return store.checkpoint_states[store.current_epoch_observed_justified_checkpoint]
```

#### LMD-GHOST helpers

##### `get_block_support_between_slots`

*Notes:*

Due to the algorithm logic, maximum distance between `balance_source` and
shuffling is two epochs which is less than `MAX_SEED_LOOKAHEAD` and thus the
balances are always consistent with the shuffling.

```python
def get_block_support_between_slots(
    store: Store,
    balance_source: BeaconState,
    block_root: Root,
    start_slot: Slot,
    end_slot: Slot,
) -> Gwei:
    """
    Return support of the block between ``start_slot`` and ``end_slot`` (inclusive of both).
    """
    participants: set[ValidatorIndex] = set()
    for slot in range(start_slot, end_slot + 1):
        participants.update(get_slot_committee(store, Slot(slot)))

    return Gwei(
        sum(
            balance_source.validators[i].effective_balance
            for i in participants
            # Check that validator is in the balance source set,
            # has voted in the support of the block
            # and hasn't been slashed.
            if (
                i < len(balance_source.validators)

>                 and i in store.latest_messages

Author: To the original question on votes in future epochs (after `epoch(end_slot)`), I think the situation is even less likely to happen.

>                 and i in store.latest_messages

Reviewer 3: I don't know the intent. The options seem to be:

1. Count _any_ late vote, regardless of the slot in which it took place (including from validators in other committees who have a stale view, e.g., assigned after end_slot within the same epoch)
2. **[current impl]** Count late votes within the given slot range, plus late votes from future epochs by the same validators (any slot)
3. **[current docstring]** Count late votes only within the given slot range, if there was another vote in future epoch ignore it (as we don't know anymore if there was a vote within the slot range, or if the assignment during that epoch was missed)

Security-wise, I guess (3) > (2) > (1), as it increases the conservatively undercounted support in unclear situations.

For the extra tracking for (3), one could track separate set per epoch inside the function, then just check store.latest_messages[i].epoch if it matches. Or, alternatively, track LatestMessage.slot instead of LatestMessage.epoch which client implementations may already do for other reasons.

If the goal is to keep the impl at (2), the docs should be updated to match the implementation. It is unclear to me where "unlikely" is relevant and where it isn't (some of this can also be set up intentionally by an attacker). There should either be a test that checks this edge case and verifies that implementations correctly overcount in that situation in a consistent way (as in, impl matches spec), or alternatively a comment that documents that it is okay to conservatively undercount if an impl has more information to do this more accurately.

(1) has the advantage that it is simple and doesn't try to pretend that one can compute an accurate support score. If it is good enough as an approximation, could be a legit option.

>                 and i in store.latest_messages

Reviewer 3: So, to document the current reality, the function effectively counts _any_ vote for `block_root` from validators that happened to be assigned during the gap til the next block on the canonical branch, regardless of whether the vote actually happened during the gap. Whether this is expected behaviour, not sure, the docs suggest different intent.

        return Gwei(0)
```

##### `get_support_discount`

```python
def get_support_discount(store: Store, balance_source: BeaconState, block_root: Root) -> Gwei:
    """
    Return weight that can be discounted during the safety threshold computation for the block.
    """

    # Empty slot support discount
    return compute_empty_slot_support_discount(store, balance_source, block_root)
```

##### `compute_safety_threshold`

```python
def compute_safety_threshold(store: Store, block_root: Root, balance_source: BeaconState) -> bool:
    """
    Compute the LMD_GHOST safety threshold for ``block_root``.
    """
    current_slot = get_current_slot(store)
    block = store.blocks[block_root]
    parent_block = store.blocks[block.parent_root]

    proposer_score = compute_proposer_score(balance_source)
    total_active_balance = get_total_active_balance(balance_source)
    maximum_support = estimate_committee_weight_between_slots(
        total_active_balance, Slot(parent_block.slot + 1), Slot(current_slot - 1)
    )
    support_discount = get_support_discount(store, balance_source, block_root)
    adversarial_weight = get_adversarial_weight(store, balance_source, block_root)

>     # The subtraction is safe as maximum_support is always greater than support_discount

Reviewer 3: An even simpler scenario would be a unevenly distributed validator set, e.g., with validator 0 as 2048 ETH, and validators 1-7 as 32 ETH.

- maximum_support = (2048 + 7 * 32)/8 = 2272/8 = 284 ETH
- proposer_score = 2272/8 * 40% = 113.6 ETH
- support_discount (assuming val 0 voted) = 2048 - (2272/8 * 25%) = 2048 - 71 = 1977 ETH

284 + 113.6 - 1977 --> underflow

    participants: set[ValidatorIndex] = set()
    for i in range(committees_count):
        participants.update(get_beacon_committee(shuffling_source, slot, CommitteeIndex(i)))
    return participants
```

##### `get_pulled_up_head_state`

```python
def get_pulled_up_head_state(store: Store) -> BeaconState:
    """
    Return the state of the head pulled up to the current epoch if needed.
    """
    head = get_head(store)
    head_state = store.block_states[head]
    if get_current_epoch(head_state) < get_current_store_epoch(store):
        pulled_up_state = copy(head_state)
        process_slots(pulled_up_state, compute_start_slot_at_epoch(get_current_store_epoch(store)))
        return pulled_up_state
    else:
        return head_state
```

##### `get_previous_balance_source`

*Notes:*

Reconfirmation is the only place where previous balance source is used.
Implementations MAY switch to the current epoch balance source after they run
reconfirmation and before the logic to confirm new blocks is called. In this
case implementations will need to keep only a single balance source around.

```python
def get_previous_balance_source(store: Store) -> BeaconState:

>     return store.checkpoint_states[store.previous_epoch_observed_justified_checkpoint]

Reviewer 3: Like, that's what I meant with "direct attestation", one that votes for A specifically, instead of for one of the descendants. Either in the block or as a standalone attestation.

>     return store.checkpoint_states[store.previous_epoch_observed_justified_checkpoint]

Reviewer 3: Ah, nvm. I have confused the FFG and GHOST votes. The `beacon_block_root` behaves like I describe, but what matters is indeed the `target`. And that one cannot be justified without anyone attesting specifically for it. Sorry for the noise, my bad.
